ext/hook/uhook.c

```diff
@@ -0,0 +1,331 @@
+#include <php.h>
+
+#include <zend_closures.h>
+#include <zend_generators.h>
+
+#if PHP_VERSION_ID < 80000
+#include "php7/compatibility.h"
+#include "php7/configuration.h"
+#else
+#include "php8/configuration.h"
+#endif
+
+#include "uhook_arginfo.h"
+
+#include <hook/hook.h>
+
+#include "uhook.h"
+
+extern void (*profiling_interrupt_function)(zend_execute_data *);
+
+static inline zval *ddtrace_hookdata_property_id(zend_object *hookdata) {
+    return OBJ_PROP_NUM(hookdata, 0);
+}
+static inline zval *ddtrace_hookdata_property_args(zend_object *hookdata) {
+    return OBJ_PROP_NUM(hookdata, 1);
+}
+static inline zval *ddtrace_hookdata_property_returned(zend_object *hookdata) {
+    return OBJ_PROP_NUM(hookdata, 2);
+}
+static inline zval *ddtrace_hookdata_property_exception(zend_object *hookdata) {
+    return OBJ_PROP_NUM(hookdata, 3);
+}
+
+zend_class_entry *ddtrace_hook_data_ce;
+
+static __thread HashTable dd_active_hooks;
+
+typedef struct {
+    zend_object *begin;
+    zend_object *end;
+    bool running;
+    zend_long id;
+
+    zend_function *resolved;
+    zend_string *scope;
+    zend_string *function;
+} dd_uhook_def;
+
+typedef struct {
+    zend_object *hook_data;
+} dd_uhook_dynamic;
+
+HashTable *dd_uhook_collect_args(zend_execute_data *execute_data) {
+    uint32_t num_args = EX_NUM_ARGS();
+
+    HashTable *ht = emalloc(sizeof(*ht));
+    zend_hash_init(ht, num_args, NULL, ZVAL_PTR_DTOR, 0);
+
+    if (!num_args) {
+        return ht;
+    }
+
+    zval *p = EX_VAR_NUM(0);
+    zend_function *func = EX(func);
+    ht->nNumOfElements = num_args;
+
+    zend_hash_real_init(ht, 1);
+    ZEND_HASH_FILL_PACKED(ht) {
+        if (EX(func)->type == ZEND_USER_FUNCTION) {
+            uint32_t first_extra_arg = MIN(num_args, func->op_array.num_args);
+
+            for (zval *end = p + first_extra_arg; p < end; ++p) {
+                if (Z_OPT_REFCOUNTED_P(p)) {
+                    Z_ADDREF_P(p);
+                }
+                ZEND_HASH_FILL_ADD(p);
+            }
+
+            p = EX_VAR_NUM(func->op_array.last_var + func->op_array.T);
+            num_args -= first_extra_arg;
+        }
+
+        for (zval *end = p + num_args; p < end; ++p) {
+            if (Z_OPT_REFCOUNTED_P(p)) {
+                Z_ADDREF_P(p);
+            }
+            ZEND_HASH_FILL_ADD(p);
+        }
+    }
+    ZEND_HASH_FILL_END();
+
+    return ht;
+}
+
+static void dd_uhook_call_hook(zend_execute_data *execute_data, zend_object *closure, zend_object *hook_data) {
+    zval closure_zv, hook_data_zv;
+    ZVAL_OBJ(&closure_zv, closure);
+    ZVAL_OBJ(&hook_data_zv, hook_data);
+
+    bool has_this = getThis() != NULL;
+    zval rv;
+    zai_symbol_call(has_this ? ZAI_SYMBOL_SCOPE_OBJECT : ZAI_SYMBOL_SCOPE_GLOBAL, has_this ? &EX(This) : NULL, ZAI_SYMBOL_FUNCTION_CLOSURE, &closure_zv, &rv, 1, &hook_data_zv);
```

Comment:
Clang format must not be inspecting this directory, because this is 177 columns.

Reply:
[Clang format must have been disabled, because it's not running at all.](https://github.com/DataDog/dd-trace-php/pull/1619)

Basically, we (I and Joe) want to use code style which fits better what we're used to.